File: backend/app/models/experiment.py
from typing import Optional, List, Dict, Any
from datetime import datetime, timezone
from pydantic import Field
from beanie import Document
from uuid import UUID, uuid4
from app.db.enums import ExperimentStatus

class Experiment(Document):
    id: UUID = Field(default_factory=uuid4)
    tenant_id: UUID
    organization_id: Optional[UUID] = None
    project_id: UUID
    owner_id: Optional[UUID] = None
    reviewer_id: Optional[UUID] = None
    protocol_id: Optional[UUID] = None
    experiment_code: str
    title: str
    objective: Optional[str] = None
    hypothesis: Optional[str] = None
    description: Optional[str] = None
    status: ExperimentStatus = ExperimentStatus.DRAFT
    priority: str = "MEDIUM"
    metadata_json: Dict[str, Any] = Field(default_factory=dict)
    is_archived: bool = False
    is_deleted: bool = False
    start_date: Optional[datetime] = None
    planned_end_date: Optional[datetime] = None
    completed_date: Optional[datetime] = None
    reviewed_date: Optional[datetime] = None
    archived_at: Optional[datetime] = None
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    class Settings:
        name = "experiments"

class ExperimentCollaborator(Document):
    id: UUID = Field(default_factory=uuid4)
    experiment_id: UUID
    user_id: UUID
    role: str = "collaborator"
    tenant_id: UUID
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    class Settings:
        name = "experiment_collaborators"

class ExperimentAttachment(Document):
    id: UUID = Field(default_factory=uuid4)
    experiment_id: UUID
    file_name: str
    file_path: str
    file_size: int = 0
    mime_type: Optional[str] = None
    uploaded_by: Optional[UUID] = None
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    class Settings:
        name = "experiment_attachments"


class ExperimentQAComment(Document):
    id: UUID = Field(default_factory=uuid4)
    experiment_id: UUID
    tenant_id: UUID
    author_id: UUID
    author_name: str
    author_role: str = "QA"
    section_id: str
    section_title: Optional[str] = None
    target_text: Optional[str] = None
    comment: str
    category: str = "QA_REVIEW"
    status: str = "open"  # 'open' | 'resolved'
    resolved_by: Optional[str] = None
    resolved_at: Optional[datetime] = None
    resolution_note: Optional[str] = None
    replies: List[Dict[str, Any]] = Field(default_factory=list)
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    class Settings:
        name = "experiment_qa_comments"

File: frontend/src/types/experiment.ts
export interface ExperimentCollaborator {
  user_id: string;
  role: string;
  id: string;
  added_at: string;
  added_by?: string | null;
}

export interface ExperimentAttachment {
  id: string;
  file_name: string;
  file_path: string;
  file_size: number;
  mime_type?: string | null;
  uploaded_by?: string | null;
  created_at: string;
}

export interface Experiment {
  experiment_code: string;
  title: string;
  objective?: string | null;
  hypothesis?: string | null;
  description?: string | null;
  status: string;
  priority: string;
  protocol_id?: string | null;
  start_date?: string | null;
  planned_end_date?: string | null;
  metadata_json: Record<string, any>;
  id: string;
  tenant_id: string;
  organization_id: string;
  project_id: string;
  owner_id?: string | null;
  reviewer_id?: string | null;
  completed_date?: string | null;
  reviewed_date?: string | null;
  is_archived: boolean;
  archived_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ExperimentDetail extends Experiment {
  collaborators: ExperimentCollaborator[];
  attachments: ExperimentAttachment[];
}

export interface ExperimentCreate {
  experiment_code: string;
  title: string;
  objective?: string | null;
  hypothesis?: string | null;
  description?: string | null;
  status?: string;
  priority?: string;
  protocol_id?: string | null;
  start_date?: string | null;
  planned_end_date?: string | null;
  metadata_json?: Record<string, any>;
  project_id?: string | null;
  organization_id?: string | null;
  tenant_id?: string | null;
  owner_id?: string | null;
  reviewer_id?: string | null;
}

export interface ExperimentUpdate {
  title?: string | null;
  objective?: string | null;
  hypothesis?: string | null;
  description?: string | null;
  status?: string | null;
  priority?: string | null;
  protocol_id?: string | null;
  start_date?: string | null;
  planned_end_date?: string | null;
  completed_date?: string | null;
  reviewed_date?: string | null;
  reviewer_id?: string | null;
  metadata_json?: Record<string, any> | null;
}

export interface ExperimentListResponse {
  items: Experiment[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}
